android/app/src/main/java/com/dismoi/scout/Floating/Layout/Bubble.java

```diff
@@ -0,0 +1,211 @@
+          playAnimationClickDown();
+          lastTouchDown = System.currentTimeMillis();
+          updateSize();
+          animator.stop();
+          break;
+        case MotionEvent.ACTION_MOVE:
+          int x = initialX + (int)(event.getRawX() - initialTouchX);
+          int y = initialY + (int)(event.getRawY() - initialTouchY);
+          getViewParams().x = x;
+          getViewParams().y = y;
+          getWindowManager().updateViewLayout(this, getViewParams());
+          if (getLayoutCoordinator() != null) {
+              getLayoutCoordinator().notifyBubblePositionChanged(this, x, y);
+          }
+          break;
+        case MotionEvent.ACTION_UP:
+          goToWall();
+          if (getLayoutCoordinator() != null) {
+              getLayoutCoordinator().notifyBubbleRelease(this);
+              playAnimationClickUp();
+          }
+          if (System.currentTimeMillis() - lastTouchDown < TOUCH_TIME_THRESHOLD) {
+              if (onBubbleClickListener != null) {
+                  onBubbleClickListener.onBubbleClick(this);
+              }
+          }
+          break;
+      }
+    }
+    return super.onTouchEvent(event);
+  }
+
+  private void playAnimation() {
+    if (!isInEditMode()) {
+      AnimatorSet animator = (AnimatorSet) AnimatorInflater
+              .loadAnimator(getContext(), R.animator.bubble_shown_animator);
+      animator.setTarget(this);
+      animator.start();
+    }
+  }
+
+  private void playAnimationClickDown() {
+    if (!isInEditMode()) {
+      AnimatorSet animator = (AnimatorSet) AnimatorInflater
+              .loadAnimator(getContext(), R.animator.bubble_down_click_animator);
+      animator.setTarget(this);
+      animator.start();
+    }
+  }
+
+  private void playAnimationClickUp() {
+    if (!isInEditMode()) {
+      AnimatorSet animator = (AnimatorSet) AnimatorInflater
+              .loadAnimator(getContext(), R.animator.bubble_up_click_animator);
+      animator.setTarget(this);
+      animator.start();
+    }
+  }
+
+  private void updateSize() {
+    DisplayMetrics metrics = new DisplayMetrics();
+    windowManager.getDefaultDisplay().getMetrics(metrics);
+    Display display = getWindowManager().getDefaultDisplay();
+    Point size = new Point();
+    display.getSize(size);
+    width = (size.x - this.getWidth());
+  }
+
+  public interface OnBubbleRemoveListener {
+    void onBubbleRemoved(Bubble bubble);
+  }
+
+  public interface OnBubbleClickListener {
+    void onBubbleClick(Bubble bubble);
+  }
+
+  public void goToWall() {
+    if (shouldStickToWall){
+      int middle = width / 2;
+      float nearestXWall = getViewParams().x >= middle ? width : 0;
+      animator.start(nearestXWall, getViewParams().y);
+    }
+  }
+
+  private void move(float deltaX, float deltaY) {
+    getViewParams().x += deltaX;
+    getViewParams().y += deltaY;
+    windowManager.updateViewLayout(this, getViewParams());
+  }
+
+
+  private class MoveAnimator implements Runnable {
```

Comment:
Should this class be declared in its own file?

Reply:
Surprisingly inner classes seems acceptable in the java community. I'm not at all a java (did it only during my school years) and OOP master so please feel free to show me some suggestions.

https://stackoverflow.com/questions/19825956/java-inner-class-access-and-best-practices

Reply:
It kind of makes sense :thinking: